Yosemite/Yosemite/Tools/Stats/TopEarnerStatsCachingFetcher.swift

```diff
@@ -0,0 +1,90 @@
+                                                                                       latestDateToInclude: latestDateToInclude,
+                                                                                       quantity: quantity)
+            let cacheEntry = CodableCacheEntry(value: ProductsReportRequestResponse(items: response), timeToLive: Constants.cacheTimeToLive)
+            cache.save(cacheEntry, forKey: cacheKey)
+        }
+
+        return convertProductsReportIntoTopEarners(siteID: siteID,
+                                                   granularity: timeRange.topEarnerStatsGranularity,
+                                                   date: latestDateToInclude,
+                                                   productsReport: response,
+                                                   quantity: quantity)
+    }
+}
+
+private extension TopEarnerStatsCachingFetcher {
+    /// Converts the `[ProductsReportItem]` list in a Products analytics report into `TopEarnerStats`
+    ///
+    func convertProductsReportIntoTopEarners(siteID: Int64,
+                                             granularity: StatGranularity,
+                                             date: Date,
+                                             productsReport: [ProductsReportItem],
+                                             quantity: Int) -> TopEarnerStats {
+        let statsDate = StatsStoreV4.buildDateString(from: date, with: granularity)
+        let statsItems = productsReport.map { product in
+            TopEarnerStatsItem(productID: product.productID,
+                               productName: product.productName,
+                               quantity: product.quantity,
+                               total: product.total,
+                               currency: "", // TODO: Remove currency https://github.com/woocommerce/woocommerce-ios/issues/2549
+                               imageUrl: product.imageUrl)
+        }
+        return TopEarnerStats(siteID: siteID, date: statsDate, granularity: granularity, limit: quantity.description, items: statsItems)
+    }
+
+    func requestCacheKey(siteID: Int64,
+                         timeZone: TimeZone,
+                         earliestDateToInclude: Date,
+                         latestDateToInclude: Date,
+                         quantity: Int) -> String {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "dd:MM:yy"
+
+        let stringArray = [String(siteID),
+                           timeZone.identifier,
+                           dateFormatter.string(from: earliestDateToInclude),
+                           dateFormatter.string(from: latestDateToInclude),
+                           String(quantity)]
+
+        return stringArray.joined(separator: "-")
+    }
+}
+
+private enum Constants {
+    static let cacheTimeToLive: TimeInterval = 60*30
```

Comment:
Isn't it 30 minutes for caching just too short? I can imagine that but for some date ranges the data never changes, isn't it?

Reply:
I agree that at least for some date ranges we could cache the data much longer. If the end date for the date range is already in the past when we request the data, we could cache much more aggressively because the data is less likely to change.